gradio/helpers.py

```diff
@@ -0,0 +1,780 @@
+            except StopIteration:
+                self.iterables.pop()
+                raise StopIteration
+        else:
+            return self
+
+    def __call__(
+        self,
+        progress: float | Tuple[int, int | None] | None,
+        desc: str | None = None,
+        total: float | None = None,
+        unit: str = "steps",
+        _tqdm=None,
+    ):
+        """
+        Updates progress tracker with progress and message text.
+        Parameters:
+            progress: If float, should be between 0 and 1 representing completion. If Tuple, first number represents steps completed, and second value represents total steps or None if unknown. If None, hides progress bar.
+            desc: description to display.
+        """
+        if self._active:
+            self._callback(
+                event_id=self._event_id,
+                iterables=self.iterables
+                + [TrackedIterable(None, 0, total, desc, unit, _tqdm, progress)],
+            )
+        else:
+            return progress
+
+    def track(
+        self,
+        iterable: Iterable | None,
+        desc: str = None,
+        total: float = None,
+        unit: str = "steps",
+        _tqdm=None,
+        *args,
+        **kwargs,
+    ):
+        """
+        Attaches progress tracker to iterable.
+        Parameters:
+            iterable: iterable to attach progress tracker to.
+            desc: description to display.
+
+        """
+        if iterable is None:
+            new_iterable = TrackedIterable(None, 0, total, desc, unit, _tqdm)
+            self.iterables.append(new_iterable)
+            self._callback(event_id=self._event_id, iterables=self.iterables)
+            return
+        length = len(iterable) if hasattr(iterable, "__len__") else None
+        self.iterables.append(
+            TrackedIterable(iter(iterable), 0, length, desc, unit, _tqdm)
+        )
+        return self
+
+    def update(self, n=1):
+        """
+        Increases latest iterable with specified number of steps.
+        Parameters:
+            n: number of steps completed.
+        """
+        if self._active and len(self.iterables) > 0:
+            current_iterable = self.iterables[-1]
+            current_iterable.index += n
+            self._callback(
+                event_id=self._event_id,
+                iterables=self.iterables,
+            )
+        else:
+            return
+
+    def close(self, _tqdm):
+        """
+        Removes iterable with given _tqdm.
+        """
+        if self._active:
+            for i in range(len(self.iterables)):
+                if id(self.iterables[i]._tqdm) == id(_tqdm):
+                    self.iterables.pop(i)
+                    break
+            self._callback(
+                event_id=self._event_id,
+                iterables=self.iterables,
+            )
+        else:
+            return
+
+
+def create_tracker(block_parent, event_id, fn, track_tqdm):
```

Comment:
Consider renaming `block_parent` as `root_blocks`. `block_parent` sounds like the immediate parent and we use `root_blocks` in other parts of the codebase